lib/elixir/lib/code/typespec.ex

```diff
@@ -276,10 +276,20 @@ defmodule Code.Typespec do
     {struct, fields} = Keyword.pop(fields, :__struct__)
     map = {:%{}, [line: line], fields}
 
-    if struct do
-      {:%, [line: line], [struct, map]}
-    else
-      map
+    case struct do
```

Comment:
I think a better fix would be to check this:

    if is_atom(struct) and struct != nil do

After all, that's the only case we want to rewrite it to `%Foo{}`. If struct is anything else, we should probably keep its original format. For example, `%{__struct__: atom()}` should not be rewritten to `%atom(){...}` either. WDYT?

Reply:
true.